core/server/master/src/main/java/alluxio/master/file/cmdmanager/FileSystemCommandManager.java

```diff
@@ -0,0 +1,371 @@
+      CommandStatus commandStatus = mStatusMap.get(commandId);
+      commandStatus.updateOptions(options);
+    }
+    return false;
+  }
+
+  /**
+   * Schedule a command to run.
+   * @param commandInfo command meta information
+   * @throws ResourceExhaustedException throw ResourceExhaustedException
+   * @throws InterruptedException throw InterruptedException
+   */
+  public void schedule(CommandInfo commandInfo)
+          throws ResourceExhaustedException, InterruptedException {
+    long commandId = commandInfo.getId();
+    CommandStatus commandStatus = new CommandStatus(commandInfo.getId(),
+            commandInfo.getPath(), commandInfo.getCmdOptions());
+    mScheduler.schedule(commandStatus);
+    mStatusMap.put(commandId, commandStatus);
+  }
+
+  /**
+   * Get status for a command.
+   * @param id command id
+   * @return running status
+   */
+  public Status getStatus(Long id) {
+    return null;
+  }
+
+  /**
+   * Stop a command for running.
+   * @param id command id
+   */
+  public void stop(Long id) {
+  }
+
+  /**
+   * Close the command manager.
+   * @throws IOException IOException
+   */
+  public void close() throws IOException {
+  }
+
+  @Override
+  public Journaled getDelegate() {
+    return mCmdEntry;
+  }
+
+  static class Scheduler {
+    private static final int CAPACITY = 100;
+    private static final long TIMEOUT = 100;
+    private final ExecutorService mExecutorService = Executors.newSingleThreadExecutor();
+    private final BlockingQueue<CommandStatus> mCommandQueue = new LinkedBlockingQueue<>(CAPACITY);
+    private int mCurrentSize = 0;
+    private final Client mClient;
+    private long mTimeoutCounts;
+    private long mRTCounts;
+
+    public Scheduler(Client client) {
+      if (client == null) {
+        mClient = new Client();
+      } else {
+        mClient = client;
+      }
+      mTimeoutCounts = 0;
+      mRTCounts = 0;
+    }
+
+    void schedule(CommandStatus commandStatus)
+            throws ResourceExhaustedException, InterruptedException {
+      if (mCurrentSize == CAPACITY) {
+        throw new ResourceExhaustedException(
+                "Insufficient capacity to enqueue tasks!");
+      }
+
+      boolean offered = mCommandQueue.offer(commandStatus, TIMEOUT, TimeUnit.MILLISECONDS);
+      if (offered) {
+        mCurrentSize++;
+      } else {
+        LOG.warn("Cannot enqueue commandStatus to the queue, may lose track on this command!"
+                + commandStatus.getDetailedInfo());
+      }
+
+      mExecutorService.submit(() -> {
+        try {
+          mClient.runCommand(commandStatus);
+        } catch (AlluxioRuntimeException e) {
+          mRTCounts++;
+          handleErrorOnStatuses(); // handle based on status
+        } catch (TimeoutException e) {
+          mTimeoutCounts++;
+          // add retry and handle timeout caused by checking available workers
+        }
+      });
+    }
+
+    void start() {
+      Thread thread = new Thread(() -> {
```

Comment:
What is this thread for? Removing finished commands? Can't we just do this once a command is finished in the executor thread?